File: src/test/java/team/yi/rsql/querydsl/test/javatest/CustomFieldTypeHandlerTest.java
package team.yi.rsql.querydsl.test.javatest;

import org.junit.jupiter.api.Test;
import team.yi.rsql.querydsl.QuerydslRsql;
import team.yi.rsql.querydsl.RsqlConfig;
import team.yi.rsql.querydsl.exception.RsqlException;
import team.yi.rsql.querydsl.model.Car;
import team.yi.rsql.querydsl.operator.RsqlOperator;
import team.yi.rsql.querydsl.test.BaseRsqlTest;
import team.yi.rsql.querydsl.test.javatest.handler.CustomFieldTypeHandler;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

@SuppressWarnings("SpellCheckingInspection")
public class CustomFieldTypeHandlerTest extends BaseRsqlTest {
    @Test
    public void shouldReadRsqlConfigWithOperator() {
        RsqlConfig rsqlConfig = new RsqlConfig.Builder(this.entityManager)
            .operator(new RsqlOperator("=customnotempty="))
            .javaFieldTypeHandler(CustomFieldTypeHandler.class)
            .build();
        QuerydslRsql<Car> rsql = new QuerydslRsql.Builder<Car>(rsqlConfig)
            .from("Car")
            .where("description=customnotempty=''")
            .build();
        List<?> cars = rsql.buildJPAQuery().fetch();

        assertNotNull(cars, "result is null");
        assertFalse(cars.isEmpty(), "Can't read config with custom operator");
        assertEquals(50, cars.size(), "Can't read config correctly with custom operator");
    }

    @Test
    public void shouldNotFindCustomOperator() {
        assertThrows(RsqlException.class, () -> {
            RsqlConfig rsqlConfig = new RsqlConfig.Builder(this.entityManager)
                .javaFieldTypeHandler(CustomFieldTypeHandler.class)
                .build();
            QuerydslRsql<Car> rsql = new QuerydslRsql.Builder<Car>(rsqlConfig)
                .from("Car")
                .where("description=customnotempty=''")
                .build();

            rsql.buildJPAQuery().fetch();
        });
    }

    @Test
    public void shouldReadRsqlConfigWithFieldTypeHandler() {
        RsqlConfig rsqlConfig = new RsqlConfig.Builder(this.entityManager)
            .javaFieldTypeHandler(CustomFieldTypeHandler.class)
            .build();
        QuerydslRsql<Car> rsql = new QuerydslRsql.Builder<Car>(rsqlConfig)
            .from("Car")
            .where("description=notempty=''")
            .build();
        List<?> cars = rsql.buildJPAQuery().fetch();

        assertNotNull(cars, "result is null");
        assertFalse(cars.isEmpty(), "Can't read config with custom field type handler");
        assertEquals(50, cars.size(), "Can't read config correctly with custom field type handler");
    }
}

File: src/test/kotlin/team/yi/rsql/querydsl/test/kotlintest/CustomFieldTypeHandlerTest.kt
package team.yi.rsql.querydsl.test.kotlintest

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import team.yi.rsql.querydsl.*
import team.yi.rsql.querydsl.exception.RsqlException
import team.yi.rsql.querydsl.model.Car
import team.yi.rsql.querydsl.operator.RsqlOperator
import team.yi.rsql.querydsl.test.BaseRsqlTest
import team.yi.rsql.querydsl.test.kotlintest.handler.CustomFieldTypeHandler

@Suppress("SpellCheckingInspection")
class CustomFieldTypeHandlerTest : BaseRsqlTest() {
    @Test
    fun shouldReadRsqlConfigWithOperator() {
        val rsqlConfig = RsqlConfig.Builder(entityManager)
            .operator(RsqlOperator("=customnotempty="))
            .fieldTypeHandler(CustomFieldTypeHandler::class.java)
            .build()
        val rsql = QuerydslRsql.Builder<Car>(rsqlConfig)
            .from("Car")
            .where("description=customnotempty=''")
            .build()
        val cars = rsql.buildJPAQuery().fetch()

        assertNotNull(cars, "result is null")

        cars?.let {
            assertFalse(cars.isEmpty(), "Can't read config with custom operator")
            assertEquals(50, cars.size, "Can't read config correctly with custom operator")
        }
    }

    @Test
    fun shouldNotFindCustomOperator() {
        assertThrows(RsqlException::class.java) {
            val rsqlConfig = RsqlConfig.Builder(entityManager)
                .fieldTypeHandler(CustomFieldTypeHandler::class.java)
                .build()
            val rsql = QuerydslRsql.Builder<Car>(rsqlConfig)
                .from("Car")
                .where("description=customnotempty=''")
                .build()

            rsql.buildJPAQuery().fetch()
        }
    }

    @Test
    fun shouldReadRsqlConfigWithFieldTypeHandler() {
        val rsqlConfig = RsqlConfig.Builder(entityManager)
            .fieldTypeHandler(CustomFieldTypeHandler::class.java)
            .build()
        val rsql = QuerydslRsql.Builder<Car>(rsqlConfig)
            .from("Car")
            .where("description=notempty=''")
            .build()
        val cars = rsql.buildJPAQuery().fetch()

        assertNotNull(cars, "result is null")

        cars?.let {
            assertFalse(cars.isEmpty(), "Can't read config with custom field type handler")
            assertEquals(50, cars.size, "Can't read config correctly with custom field type handler")
        }
    }
}
